### PR TITLE
Suggestion: Added support for Patchouli Books

### PR DESCRIPTION
**Pull request by SimonFlapse**
* This allows the creating of in-game books for documentations
* Provided example with Block Breaker and Block Placer
* Replaces old Tech Reborn Manual

## Notes
This PR is a concept of using the Patchouli mod for adding in-game books often used by other mods for documentation.

Let me know what you think of this idea

### Example screenshots

image
image
image
image

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -318,6 +328,7 @@ task renameCrowdin(type: Copy, dependsOn: ['crowdin', 'cleanCrowdin']){
 import groovy.json.JsonOutput
 import groovy.json.JsonSlurper
 import groovy.util.XmlSlurper
+import org.kohsuke.github.GHReleaseBuilder
```
Comment (modmuss50) [inline build.gradle:331]: Please dont make unrelated changes.

File: src/main/java/techreborn/items/ManualItem.java [hunk 2]
```diff
@@ -45,9 +47,15 @@ public ManualItem() {
 	@Override
 	public TypedActionResult<ItemStack> use(final World world, final PlayerEntity player, final Hand hand) {
 		if (player instanceof ServerPlayerEntity serverPlayerEntity) {
-			NetworkManager.sendToPlayer(ClientboundPackets.createPacketOpenManual(), serverPlayerEntity);
+			PatchouliAPI.get().openBookGUI(serverPlayerEntity, new Identifier("techreborn:techreborn"));
```
Comment (modmuss50) [inline src/main/java/techreborn/items/ManualItem.java:50]: This will crash when patchouli is not present.

This should send a packet as I expect `openBookGUI` is client only, this breaks the refactors I made to help work towards: https://github.com/TechReborn/TechReborn/pull/2846

File: src/main/resources/assets/techreborn/lang/en_us.json [hunk 3]
```diff
@@ -1106,5 +1107,13 @@
...
-  "gui.techreborn.block.progress.paused": "Idle (%s)"
+  "gui.techreborn.block.progress.paused": "Idle (%s)",
+
+  "_comment27": "Patchouli Book",
+  "patchouli.techreborn.guide.name": "Tech Reborn Manual",
+  "patchouli.techreborn.landing_text": "Welcome to the world of Tech Reborn.$(br2)Check out our online $(l:https://wiki.techreborn.ovh/doku.php)Wiki$()$(br2)You can also join our $(l:https://discord.gg/teamreborn)Discord!$()$(br2)If you have feedback, suggestions, issues or want to contribute visit our $(l:https://github.com/TechReborn/TechReborn/issues)GitHub$()$(br2)We hope you enjoy playing!",
+
+  "_comment28": "Guide Book Machines",
+  "patchouli.techreborn.machine.block_breaker": "The block breaker is a simple machine. It allows you to break a block from the location in front of it, to its inventory. This action costs energy",
+  "patchouli.techreborn.machine.block_placer": "The block placer is a simple machine. It allows you to place a block from its inventory, to the location in front of it. This action costs energy"
```
Comment (modmuss50) [inline src/main/resources/assets/techreborn/lang/en_us.json:1118]: Im unsure if this is the best way to document TR, the wiki is the go-to place for documentation, im unsure if we want to fragment it even more?

**Review by modmuss50 — CHANGES_REQUESTED** [review]
Some issues with this, and im unsure if this is something we even need/want as it will fragment the documentation across multiplue places.

File: build.gradle (hunk 1 above)
Comment (SimonFlapse) [reply build.gradle:331]: Hmm weird I do not remember adding that.

File: src/main/java/techreborn/items/ManualItem.java (hunk 2 above)
Comment (SimonFlapse) [reply src/main/java/techreborn/items/ManualItem.java:50]: I'll look into fixing this if this concept is something we want

**Review by Ayutac — COMMENTED** [review]
Please close this PR. Only a completed Patchouli book should be added. This is already in the works. I added you on the discord thread.

**Comment by SimonFlapse** [comment]
> Some issues with this, and im unsure if this is something we even need/want as it will fragment the documentation across multiplue places.

I have experienced multiple people being very annoyed trying to craft the book, for it to just not offer any information. It feels like people are starting to get accustomed to a fully fledged reference guide in game. Furthermore this will closely link the documentation to the actual code (Wikis often become outdated, (or updated beyond the version a player is using) but an included book is expected to be up-to-date for the version it is released with.

I should clarify, I do not expect this to be merged in its current state, I'll convert it to a draft. If this is something that is interesting, this would serve as a starting point.

File: src/main/resources/assets/techreborn/lang/en_us.json (hunk 3 above)
Comment (Ayutac) [reply src/main/resources/assets/techreborn/lang/en_us.json:1118]: This is totally wrong. Patchouli is worked on per translation. Ie rather than adding translation keys, you add a complete book in the correct language, which is explained on the patchouli book page as well.

Regarding the wiki, it is bad insofa that it is not being kept up to date. I want to do something with datagen to automatically keep the wiki and the patchouli up to date and synchronized

**Comment by Ayutac** [comment]
a clean smallscale reference implementation can be found here

https://github.com/Ayutac/time/commit/24264bd9d86e3cd2684f4a4f5146dd4f2f471cb7

File: src/main/resources/assets/techreborn/lang/en_us.json (hunk 3 above)
Comment (SimonFlapse) [reply src/main/resources/assets/techreborn/lang/en_us.json:1118]: Patchouli has two options for i18n

I opted for the optional Lang files support to allow the already working crowd source translation to work instead of having a new book for each language. (As suggested by the mod).

Comment (Ayutac) [reply src/main/resources/assets/techreborn/lang/en_us.json:1118]: Okay, I admit I didn't know that.
Nevertheless, this was makes maintaining the book much harder imo

Comment (SimonFlapse) [reply src/main/resources/assets/techreborn/lang/en_us.json:1118]: I agree, it is harder to create the book this way, but on the other hand, the alternative fragments the way translations are handled. I wanted to spark discussion.

**Comment by Ayutac** [comment]
Let's do it properly. A good patchouli guide takes takes time and work and shouldn't hang on a single person's dev branch. I opened https://github.com/TechReborn/TechReborn/tree/feature/patchouli, lets work there with real PRs instead of a single draft PR onto dev

**Comment by SimonFlapse** [comment]
> Let's do it properly. A good patchouli guide takes takes time and work and shouldn't hang on a single person's dev branch. I opened https://github.com/TechReborn/TechReborn/tree/feature/patchouli, lets work there with real PRs instead of a single draft PR onto dev

Sounds like a good idea :)
